doc/monads/readers.lean

```diff
@@ -0,0 +1,196 @@
+
+def readerFunc3 : ReaderM Environment String := do
+  let x ← readerFunc2
+  return "Result: " ++ (toString x)
+
+def main2 : IO Unit := do
+  let env ← loadEnv
+  let str := readerFunc3 env
+  IO.println str
+
+#eval main2 -- Result: 7538
+/-!
+`main2` loads the environment as before, and estabilishes the `ReaderM` state by passing `env` to
+readerFunc3.  The `do` notation you learned about in [Monads](monads.lean.md) is hiding some things
+here.  Technically `readerFunc3` is a monadic action and in order to "run" that action the `ReaderM`
+monad provides a `run` method and it is the `ReaderM` run method that takes the initial
+`Environment` state.  So you can actually write the complete code by doing this: `let str :=
+readerFunc3.run env`.
+
+> **Side note 1**: The `return` statement used above also needs some explanation.  The `return`
+statement in Lean can only be used inside a `do` block and is closely related to `pure`, but a
+little different. First the similarity is that `return` and `pure` both lift a pure value up to the
+Monad type. But `return` is a keyword so you do not need to parenthesize the expression like you do
+when using `pure`.  (Note: you can avoid parentheses when using `pure` by using the `<|` operator
+like we did above in the initial `getEnvDefault` function).  Furthermore, `return` can also cause an
+early `return` in a monadic function similar to how it can in an imperative language while `pure`
+cannot.
+
+> So technically if `return` is the last statement in a function it could be replaced with `pure <|`,
+but one could argue that `return` is still a little easier for most folks to read, just so long as
+you understand that `return` is doing more than other languages, it is also wrapping pure values in
+the monadic container type.
+
+> **Side note 2**: If the function `readerFunc3` also took some explicit arguments then you would have
+to write `(readerFunc3 args).run env` and this is a bit ugly, so Lean provides an infix operator
+`|>` that eliminiates those parens so you can write `readerFunc3 args |>.run env` and then you can
+chain multiple monadic actions like this `m1 args1 |>.run args2 |>.run args3` and this is the
+recommended style.  You will see this patten used heavily in Lean code.
+
+The `let env ← read` expression in `readerFunc1` unwraps the environment from the `ReaderM` so we
+can use it. Each type of monad might provide one or more extra functions like this, functions that
+become available only when you are in the context of that monad.
+
+Here the `readerFunc2` function uses the `bind` operator `>>=` just to show you that there are bind
+operations happening here.  The `readerFunc3` function uses the  `do` notation you learned about in
+[Monads](monads.lean.md) which hides that bind operation and can make the code look cleaner.
+
+The `do` notation with `let x ← readerFunc2` is also calling the `run` function to actually execute
```

Comment:
Ok I was confused about that, can you tell me more? The reason I was confused is that I can replace that line with `let x ← readerFunc2.run` and I get the exact same result.  The unwrapping of the readerFunc2 and assignment to `x` surely implies executing the monad to get that value somehow?

Reply:
Getting the same result does not imply that the `do` notation calls the function `run`. This is just not true. You can check by yourself by printing how the `do` notation has been elaborated by Lean.
```lean
set_option pp.notation false
#print readerFunc1
```

Reply:
Thanks for the handy tool - it helps clear things up, when I use this on readerFunc3 without any run I see 
```
bind readerFunc2 fun x => pure (HAppend.hAppend "Result: " (toString x))
```
and with the run I see
```
bind (ReaderT.run readerFunc2) fun x => pure (HAppend.hAppend "Result: " (toString x))
```

So I rewrote it like this:

> The `do` notation with `let x ← readerFunc2` is also calling the `bind` function under the covers,
so that you can access the unwrapped value `x` needed for the `toString x` conversion.

Reply:
fixed in https://github.com/leanprover/lean4/pull/1505/commits/9b24f8de09a70d3793d040aff7d217bed8be8acc

Reply:
Leo, related to the above discussion, I added this paragraph in [Stat](https://lovettsoftware.com/lean4/monads/states.lean.html#implementation), love to get your feedback on it:

> It is also interesting to see how much work the `do` and `←` notation are doing for you.  To
implement the `nextTurn` function without these you would have to write this, manually plumbing
the state all the way through:
```lean4
def nextTurnManually : StateM GameState Bool
| state =>
  let (i, gs) := chooseRandomMove |>.run state
  let (_, gs') := applyMove i  |>.run gs
  let (result, gs'') := isGameDone |>.run gs'
  (result, gs'')
```